### PR TITLE
Adding a stale workflow check to the taskvalidator and code cleanup.

### PR DESCRIPTION
**Pull request by agautam478**
**What changed?**
- Added a stale workflow check to the task validator.
- Added a delete workflow call if the workflow domain is deprecated or if the workflow is stale.
- Removed the call to task validator from the history layer as the plan is now to put it in the execution layer.

Upon detecting an error during the update process, the function call will be placed in the `execution` context to cover all workflow execution scenarios, ensuring a thorough and efficient approach to identifying and resolving corruption across the system. 

**Why?**
These changes were made to complete the taskvalidation service and call in the correct place in future.
With the addition of this final check the core service can be considered ready an usable. We can move forwarding with calling the `taskvalidation` step in the next PR.

**How did you test it?**
Uni tests.

**Potential risks**

**Release notes**

**Documentation Changes**

### CODE REVIEW

File: common/constants.go [hunk 1]
```diff
@@ -264,8 +264,3 @@ const (
 	// DefaultHistoryMaxAutoResetPoints is the default maximum number for auto reset points
 	DefaultHistoryMaxAutoResetPoints = 20
 )
-
-const (
```
Comment (taylanisikdemir) [inline common/constants.go:268]: thanks for eliminating these copies

File: common/taskvalidator/validateworkflow.go [hunk 2]
```diff
@@ -24,62 +24,93 @@
...
 	metricsClient metrics.Client
 	dc            cache.DomainCache
+	pr            persistence.Retryer
+	staleCheck    staleChecker
 }
 
-// NewWfChecker creates a new instance of Checker.
-func NewWfChecker(logger log.Logger, metrics metrics.Client, domainCache cache.DomainCache) Checker {
+func NewWfChecker(logger log.Logger, metrics metrics.Client, domainCache cache.DomainCache, pr persistence.Retryer) Checker {
+	zapLogger, _ := zap.NewProduction()
```
Comment (taylanisikdemir) [inline common/taskvalidator/validateworkflow.go:58]: we shouldn't create side loggers like this. can `NewStaleWorkflow` accept `log.Logger` instead?

File: common/taskvalidator/validateworkflow.go [hunk 3]
```diff
@@ -24,62 +24,93 @@
...
+	staleCheckInstance := invariant.NewStaleWorkflow(pr, domainCache, zapLogger)
+
+	staleCheck, _ := staleCheckInstance.(staleChecker)
```
Comment (taylanisikdemir) [inline common/taskvalidator/validateworkflow.go:61]: let's not ignore the cast result ok or not. if it is not ok this function should return an error

File: common/taskvalidator/validateworkflow.go (hunk 2 above)
Comment (agautam478) [reply common/taskvalidator/validateworkflow.go:58]: it wasn't able to but maybe with the new stalechecker it will. I will check.

File: common/constants.go (hunk 1 above)
Comment (agautam478) [reply common/constants.go:268]: Thanks for the guidance.

File: common/taskvalidator/validateworkflow.go [hunk 4]
```diff
@@ -24,62 +24,123 @@
...
+	pr            persistence.Retryer
+	staleCheck    staleChecker
 }
 
-// NewWfChecker creates a new instance of Checker.
-func NewWfChecker(logger log.Logger, metrics metrics.Client, domainCache cache.DomainCache) Checker {
+// NewWfChecker creates a new instance of a workflow validation checker.
+// It requires a logger, metrics client, domain cache, persistence retryer,
+// and a stale checker implementation to function.
+func NewWfChecker(logger *zap.Logger, metrics metrics.Client, domainCache cache.DomainCache, pr persistence.Retryer, staleCheck staleChecker) Checker {
```
Comment (taylanisikdemir) [inline common/taskvalidator/validateworkflow.go:62]: where is this going to be called?

File: common/taskvalidator/validateworkflow_test.go [hunk 5]
```diff
@@ -23,45 +23,102 @@
...
+	"go.uber.org/zap"
 
 	"github.com/uber/cadence/common/cache"
-	"github.com/uber/cadence/common/log"
 	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/persistence"
 	"github.com/uber/cadence/service/history/constants"
 )
 
+type MockStaleChecker struct {
```
Comment (taylanisikdemir) [inline common/taskvalidator/validateworkflow_test.go:40]: this doesn't need to be public

File: common/taskvalidator/validateworkflow_test.go [hunk 6]
```diff
@@ -23,45 +23,102 @@
...
-
-			// Set up the mock behavior for GetDomainByID
+			mockPersistenceRetryer := persistence.NewMockRetryer(mockCtrl)
+			mockStaleChecker := &MockStaleChecker{
+				CheckAgeFunc: func(response *persistence.GetWorkflowExecutionResponse) (bool, error) {
+					return tc.isStale, nil
+				},
+			}
+			checker := NewWfChecker(mockLogger, mockMetricsClient, mockDomainCache, mockPersistenceRetryer, mockStaleChecker)
+			checker.(*checkerImpl).staleCheck = mockStaleChecker
```
Comment (taylanisikdemir) [inline common/taskvalidator/validateworkflow_test.go:78]: you already pass `mockStaleChecker` to constructor above. why is this line needed?

File: common/taskvalidator/validateworkflow.go (hunk 2 above)
Comment (taylanisikdemir) [reply common/taskvalidator/validateworkflow.go]: I see you changed constructor to expect `*zap.Logger`. that doesn't address my concern. we should keep it consistent via `log.Logger`. let's look into changing the stale checker to accept `log.Logger` not the other way around

**Review by allenchen2244 — APPROVED** [review]
